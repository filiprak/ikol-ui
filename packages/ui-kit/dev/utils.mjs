import { existsSync } from 'fs';
import { readFile, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function useTemplateFile(filename, outputFilepath, variables) {
  const data = await readFile(path.resolve(__dirname, 'templates', filename), 'utf8');
  return useTemplate(data, outputFilepath, variables);
}

export async function useTemplate(templateStr, outputFilepath, variables) {
  const updatedTemplate = templateStr.replace(/\[\[([a-zA-Z0-9_]+)\]\]/g, (match, key) => {
    return variables[key] || match;
  });

  if (!existsSync(path.dirname(outputFilepath))) {
    await mkdir(path.dirname(outputFilepath));
  }
  await writeFile(outputFilepath, updatedTemplate);
}

export const kebabize = str => {
  return str.split('').map((letter, idx) => {
      return letter.toUpperCase() === letter
          ? `${idx !== 0 ? '-' : ''}${letter.toLowerCase()}`
          : letter;
  }).join('');
}

export const ROOT_DIR = path.resolve(__dirname, '..');
