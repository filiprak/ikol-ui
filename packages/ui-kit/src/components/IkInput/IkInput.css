.ik-input {
    position: relative;
}

.ik-input__append,
.ik-input__prepend {
    display: flex;
    align-items: center;
    justify-content: center;
    position: absolute;
    top: 0;
    bottom: 0;
    line-height: 24px;
    width: 32px;
    padding: 4px 8px;
    color: var(--on-background-color-3);
}

.ik-input__prepend {
    left: 0;
}

.ik-input__append {
    right: 0;
}

.ik-input--prepend > .ik-input__native {
    padding-left: 32px !important;
}

.ik-input--append > .ik-input__native {
    padding-right: 32px !important;
}

.ik-input:focus-within > .ik-input__prepend,
.ik-input:focus-within > .ik-input__append {
    color: var(--primary-color);
}

.ik-input--disabled > .ik-input__prepend,
.ik-input--disabled > .ik-input__append {
    opacity: 0.65;
    pointer-events: none;
}
