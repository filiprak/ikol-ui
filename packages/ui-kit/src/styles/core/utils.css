/* shadow */
.ik-shadow-0 {
    box-shadow: none;
}

.ik-shadow-1 {
    box-shadow: 0 1px 3px 0 var(--shadow-color-1);
}

.ik-shadow-2 {
    box-shadow: 0 2px 5px 0 var(--shadow-color-1);
}

.ik-shadow-3 {
    box-shadow: 0 2px 10px 0 var(--shadow-color-1);
}

.ik-shadow-4 {
    box-shadow: 0 2px 15px 0 var(--shadow-color-1);
}

.ik-shadow-5 {
    box-shadow: 0 2px 20px 0 var(--shadow-color-1);
}

.ik-shadow-6 {
    box-shadow: 0 3px 30px 0 var(--shadow-color-1);
}

/* skeleton background */
@-webkit-keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

@keyframes skeleton-loading {
    to {
        transform: translateX(100%)
    }
}

.ik-bg-skeleton {
    background: var(--skeleton-color-1);
    overflow: hidden;
    position: relative;
    user-select: none;
}

.ik-bg-skeleton:after {
    background: linear-gradient(90deg, hsla(0, 0%, 100%, 0), var(--skeleton-color-2), hsla(0, 0%, 100%, 0));
    -webkit-animation: skeleton-loading 1.5s infinite;
    animation: skeleton-loading 1.5s infinite;
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    transform: translateX(-100%);
    z-index: 1
}

.ik-cursor--pointer {
    cursor: pointer;
}

/* design utils */
.ik-design--bg-primary {
    background-color: var(--primary-color);
    color: var(--on-primary-color);
}

.ik-design--bg-success {
    background-color: var(--success-color);
    color: var(--on-success-color);
}

.ik-design--bg-error {
    background-color: var(--error-color);
    color: var(--on-error-color);
}

.ik-design--bg-accent {
    background-color: var(--accent-color);
    color: var(--on-accent-color);
}

.ik-design--bg-secondary {
    background-color: var(--secondary-color);
    color: var(--on-secondary-color);
}

.ik-design--bg-default {
    background-color: var(--surface-darken-color);
    color: var(--on-surface-darken-color-3);
}

/* height */
.ik-height--full {
    height: 100%;
}

.ik-width--full {
    width: 100%;
}
