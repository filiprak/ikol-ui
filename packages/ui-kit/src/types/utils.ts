import type { ExtractPropTypes } from 'vue';

/* eslint-disable @typescript-eslint/naming-convention */
export type IkExtractProps<T> =
    T extends new (...angs: any) => { $props: infer P; } ? NonNullable<P> : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: infer P, ...args: any) => any ? P : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkExtractPropTypes<T> = Prettify<Readonly<ExtractPropTypes<T>>>;

export type IkExtractExposed<T> =
    T extends new (...angs: any) => infer E ? E : // eslint-disable-line @typescript-eslint/no-explicit-any
    T extends (props: any, ctx: any, expose: (exposed: infer E) => any, ...args: any) => any ? NonNullable<E> : // eslint-disable-line @typescript-eslint/no-explicit-any
    {};

export type IkKeysByType<T, U> = keyof {
    [K in keyof T as T[K] extends U ? K : never]: T[K];
};

export enum IkUIDesignColor {
    primary = 'primary',
    success = 'success',
    error = 'error',
    accent = 'accent',
    default = 'default',
}

export enum IkUIElemSize {
    xs = 'xs',
    sm = 'sm',
    md = 'md',
    lg = 'lg',
}

export enum IkUIVariant {
    flat = 'flat',
    filled = 'filled',
    outline = 'outline',
    pill = 'pill',
}

export enum IkPosition {
    top = 'top',
    right = 'right',
    bottom = 'bottom',
    left = 'left',
    horizontal = 'horizontal',
    vertical = 'vertical',
}

export enum IkAlignment {
    start = 'start',
    center = 'center',
    end = 'end',
}

export type EnumT = Record<string, string> | Record<string | number, string | number>;
export type ObjValueT<T> = T[keyof T];
export type DefinedKeys<T> = { [K in keyof T]: undefined extends T[K] ? never : K }[keyof T];
export type OmitUndefined<T> = { [K in DefinedKeys<T>]: T[K] };
export type NotArray<T> = T extends any[] ? never : T; // eslint-disable-line @typescript-eslint/no-explicit-any
export type DefinedComponent = new (...args: any[]) => any; // eslint-disable-line @typescript-eslint/no-explicit-any
export type Prettify<T> = {
    [K in keyof T]: T[K];
} & {};
