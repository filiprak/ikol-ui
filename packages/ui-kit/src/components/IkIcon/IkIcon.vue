<script setup lang="ts">
import '@ui/styles';
import './IkIcon.css';
import { type PropType, h, useAttrs } from 'vue';
import { toCssUnits, getIconClasses } from '@ui/utils/helpers';
import { useRender } from '@ui/composables/render';
import type { IkUIDesignColor } from '@ui/types/utils';

const props = defineProps({
    tag: {
        type: String,
        default: 'i',
    },
    icon: {
        type: String,
    },
    size: {
        type: [Number, String],
    },
    size_px: {
        type: [Number, String],
    },
    disabled: {
        type: Boolean,
        default: false,
    },
    circle: {
        type: Boolean,
        default: false,
    },
    design: {
        type: String as PropType<`${IkUIDesignColor}`>,
    },
    variant: {
        type: String,
    },
});
const attrs = useAttrs();

useRender(() => {
    return h(props.tag, {
        class: [
            {
                'ik-icon': true,
                ['ik-icon--' + props.design]: !!props.design,
                ['ik-icon--' + props.variant]: !!props.variant,
                'ik-icon--circle': props.circle,
                'ik-icon--clickable': !props.disabled && !!attrs.onClick,
                'ik-icon--disabled': props.disabled,
                ['fa-' + props.size + 'x']: !!props.size,
            },
            getIconClasses(props.icon || ''),
        ],
        style: {
            width: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            height: props.size_px && props.circle ? toCssUnits(props.size_px, 'px') : null,
            fontSize: props.size_px ? toCssUnits(props.circle ? (Number(props.size_px) / 2) : props.size_px, 'px') : null,
        },
    });
});
</script>
