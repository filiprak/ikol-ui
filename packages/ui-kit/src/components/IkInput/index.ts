import IkInput from './IkInput.vue';
export interface IkInput extends InstanceType<typeof IkInput> { };
export { IkInput };
