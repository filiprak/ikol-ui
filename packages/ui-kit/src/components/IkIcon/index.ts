import IkIcon from './IkIcon.vue';
export interface IkIcon extends InstanceType<typeof IkIcon> { };
export { IkIcon };
