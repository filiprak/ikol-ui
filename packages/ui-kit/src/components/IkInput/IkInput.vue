<script lang="ts">
import '@ui/styles';
import './IkInput.css';
import { defineComponent, h, ref } from 'vue';
import { formatNumber, isNumeric } from '@ui/utils/helpers';
import { IkIcon } from '@ui/components/IkIcon';

export default defineComponent({
    render() {
        const input = h('input', {
            ref: 'native_el',
            class: this.computed_class,
            type: this.computed_type,
            name: this.name,
            required: this.required,
            readonly: this.readonly,
            disabled: this.disabled,
            autocomplete: this.autocomplete,
            placeholder: this.placeholder,
            min: this.min,
            max: this.max,
            minlength: this.minlength,
            maxlength: this.maxlength,
            step: this.step,
            inputmode: this.computed_inputmode,
            value: this.computed_value,
            onFocusout: this._onFocusOut,
            onInput: this._onInput,
        });

        let prepend, append;

        if (this.has_prepend) {
            prepend = h('div', { class: 'ik-input__prepend' }, [
                this.$slots.prepend?.()
                    ? this.$slots.prepend?.()
                    : h(IkIcon, {
                        icon: this.prepend_icon || this.icon,
                        ...this.prepend_on,
                    }),
            ]);
        }

        if (this.has_append) {
            append = h('div', {
                class: 'ik-input__append',
                style: this.show_append ? {} : { display: 'none' },
            }, [
                this.$slots.append?.()
                    ? this.$slots.append?.()
                    : h(IkIcon, {
                        icon: this.append_icon,
                        ...this.append_on,
                    }),
            ]);
        }

        return h('div', {
            class: {
                'ik-input': true,
                'ik-input--prepend': this.has_prepend,
                'ik-input--append': this.has_append,
                'ik-input--disabled': this.disabled,
            },
        }, [prepend, input, append]);
    },
    props: {
        modelValue: {
            type: [String, Number],
            default: '',
        },
        type: {
            type: String,
            default: 'text',
        },
        name: {
            type: String,
        },
        size: {
            type: String,
            default: 'sm',
        },
        required: {
            type: Boolean,
            default: false,
        },
        readonly: {
            type: Boolean,
            default: false,
        },
        disabled: {
            type: Boolean,
            default: false,
        },
        autocomplete: {
            type: String,
            default: 'off',
        },
        placeholder: {
            type: String,
        },
        min: {
            type: [String, Number],
        },
        max: {
            type: [String, Number],
        },
        maxlength: {
            type: [String, Number],
        },
        minlength: {
            type: [String, Number],
        },
        step: {
            type: [String, Number],
        },
        append_clear_icon: {
            type: Boolean,
            default: false,
        },
        prepend_icon: {
            type: String,
        },
        icon: {
            type: String,
        },
        append_icon: {
            type: String,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            default(props: any) {
                return props.append_clear_icon ? 'times' : undefined;
            },
        },
        decimals: {
            type: Number,
        },
        variant: {
            type: String,
            default: 'flat',
        },
        autofocus: {
            type: Boolean,
            default: false,
        },
    },
    computed: {
        computed_type() {
            return (this.type === 'decimal') ? 'text' : this.type;
        },
        computed_inputmode() {
            return (this.type === 'decimal') ? 'decimal' : undefined;
        },
        computed_class() {
            const classes: Record<string, boolean> = {
                'ik-input__native': true,
                'ik-f-input': true,
            };
            if (this.size) {
                classes['ik-f-input--size-' + this.size] = true;
            }
            classes['ik-f-input--' + this.variant] = true;
            return classes;
        },
        computed_value() {
            if (!this.hasFocus() && this.decimals !== undefined) {
                return formatNumber(this.modelValue, this.decimals);
            } else {
                return this.modelValue;
            }
        },
        has_prepend() {
            return this.prepend_icon || this.icon || this.$slots.prepend;
        },
        has_append() {
            return this.append_icon || this.$slots.append;
        },
        show_append() {
            if (this.append_clear_icon) {
                return !!this.modelValue;
            } else {
                return true;
            }
        },
        prepend_on() {
            return this.$attrs.onIkClickPrepend
                ? {
                    onClick: this.$attrs.onIkClickPrepend,
                }
                : {};
        },
        append_on() {
            const handlers = [];

            if (this.append_clear_icon) {
                handlers.push(() => {
                    this.$emit('update:modelValue', '');
                });
            }
            if (this.$attrs.onIkClickAppend) {
                handlers.push(this.$attrs.onIkClickAppend);
            }
            return { onClick: handlers };
        },
    },
    methods: {
        focus() {
            this.native_el?.focus();
        },
        hasFocus() {
            return this.native_el === document.activeElement;
        },
        moveCursorToEnd() {
            const el = this.native_el;

            if (el) {
                if (typeof el.selectionStart == 'number') {
                    el.selectionStart = el.selectionEnd = el.value.length;
                } else if (typeof (el as any).createTextRange != 'undefined') { // eslint-disable-line @typescript-eslint/no-explicit-any
                    if (!this.hasFocus()) {
                        el.focus();
                    }
                    const range = (el as any).createTextRange(); // eslint-disable-line @typescript-eslint/no-explicit-any
                    range.collapse(false);
                    range.select();
                }
            }
        },
        _setValueAsync(value: string, select = false) {
            setTimeout(() => {
                if (this.native_el) {
                    this.native_el.value = value;
                    select && this.native_el.select();
                }
            }, 0);
        },
        _onInput() {
            let value = this.native_el?.value;

            if (this.type === 'decimal') {
                if (value) {
                    const original = value;

                    value = String(value);
                    // replace commas with dots
                    value = value.replace(/,/g, '.');
                    // remove all dots except last
                    value = value.replace(/(.*)\./, (x) => { return x.replace(/\./g, '') + '.'; });
                    // replace all non-decimal characters left
                    value = value.replace(/[^\d.-]/g, '');

                    if (!isNumeric(value) && value !== '-') {
                        value = '0';
                    }
                    if (value !== original) {
                        this._setValueAsync(value, false);
                    }
                }
            }

            if ((this.type === 'number' || this.type === 'decimal') && isNumeric(value)) {
                const numeric = Number(value);

                if (this.min !== undefined && (this.min == 0 || this.min == 1) && numeric < this.min) {
                    value = String(Number(this.min));
                    this._setValueAsync(value, true);
                }
                if (this.max !== undefined && numeric > Number(this.max)) {
                    value = String(this.max);
                    this._setValueAsync(value, false);
                }
            }

            this.$emit('update:modelValue', value);
        },
        _onFocusOut() {
            if (this.decimals !== undefined) {
                const formatted = formatNumber(this.native_el?.value || '', this.decimals);

                if (this.native_el?.value !== formatted) {
                    this.$emit('update:modelValue', formatted);
                }
            }
        },
    },
    mounted() {
        if (this.autofocus) {
            setTimeout(this.focus, 0);
        }
    },
    setup() {
        return { native_el: ref<HTMLInputElement>() };
    },
});
</script>
