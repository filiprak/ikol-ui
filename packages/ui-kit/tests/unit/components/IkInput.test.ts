import type { RenderOptions } from '@ui/tests/tools/utils';
import { renderDeep, wait } from '@ui/tests/tools/utils';
import { IkInput } from '@ui/components/IkInput';
import { IkIcon } from '@ui/components/IkIcon';

describe('IkInput', () => {
    const getInstance = (options: RenderOptions<typeof IkInput> = {}) => renderDeep(IkInput, {
        attachTo: document.body,
    }, options);

    beforeEach(() => {
        (document.activeElement as HTMLElement)?.blur();
    });

    it('renders with correct default state', () => {
        const wrapper = getInstance();
        expect(wrapper.classes()).toStrictEqual([
            'ik-input',
        ]);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with correct custom state', () => {
        const wrapper = getInstance({
            props: {
                modelValue: 'Test value.',
                type: 'password',
                name: 'unit-test',
                size: 'sm',
                required: true,
                readonly: true,
                disabled: true,
                autocomplete: 'on',
                placeholder: 'Test placeholder...',
                min: 0,
                max: 15,
                maxlength: 100,
                minlength: 4,
                step: 55,
            },
        });
        expect(wrapper.find('input').element.value).toBe('Test value.');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with size', () => {
        const wrapper = getInstance({
            props: {
                size: 'lg',
            },
        });
        expect(wrapper.find('input').classes()).toContain('ik-f-input--size-lg');
    });

    it('emits input event on value change', async () => {
        const wrapper = getInstance();

        await wrapper.find('input').setValue('test');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['test'],
        ]);
    });

    it('works with focus() method', () => {
        const wrapper = getInstance();

        wrapper.vm.focus();

        expect(document.activeElement).toBe(wrapper.find('input').element);
    });

    it('works with hasFocus() method', () => {
        const wrapper = getInstance();

        expect(wrapper.vm.hasFocus()).toBe(false);

        wrapper.vm.focus();
        expect(wrapper.vm.hasFocus()).toBe(true);
    });

    it('works with decimals format', async () => {
        const wrapper = getInstance({
            props: {
                decimals: 1,
            },
        });

        expect(wrapper.find('input').element.value).toBe('');

        await wrapper.setProps({ modelValue: 5 });
        expect(wrapper.find('input').element.value).toBe('5.0');

        await wrapper.setProps({ modelValue: 'test' });
        expect(wrapper.find('input').element.value).toBe('test');
    });

    it('does not format decimals when focused', async () => {
        const wrapper = getInstance({
            props: {
                decimals: 1,
            },
        });

        wrapper.vm.focus();
        expect(wrapper.vm.hasFocus()).toBeTruthy();

        expect(wrapper.find('input').element.value).toBe('');

        await wrapper.setProps({ modelValue: 5 });
        expect(wrapper.find('input').element.value).toBe('5');

        await wrapper.setProps({ modelValue: 'test' });
        expect(wrapper.find('input').element.value).toBe('test');
    });

    it('emits event on focusout when have decimals', async () => {
        const wrapper = getInstance({
            props: {
                decimals: 3,
            },
        });

        wrapper.vm.focus();
        expect(wrapper.vm.hasFocus()).toBeTruthy();

        wrapper.find('input').element.value = '8';

        await wrapper.find('input').trigger('focusout');

        expect(wrapper.emitted('update:modelValue')).toStrictEqual([
            ['8.000'],
        ]);
        expect(wrapper.emitted('focusout')).toHaveLength(1);
        expect(wrapper.emitted('focusout')?.[0][0]).toStrictEqual(expect.any(FocusEvent));
    });

    it('renders with custom variant', () => {
        const wrapper = getInstance({
            props: {
                variant: 'flat',
            },
        });

        expect(wrapper.find('input').classes()).toContain('ik-f-input--flat');
    });

    it('renders with prepend slot', () => {
        const wrapper = getInstance({
            slots: {
                prepend: 'Unit test...',
            },
        });

        expect(wrapper.classes()).toContain('ik-input');
        expect(wrapper.classes()).toContain('ik-input--prepend');
        expect(wrapper.text()).toContain('Unit test...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with append slot', () => {
        const wrapper = getInstance({
            slots: {
                append: 'Unit test...',
            },
        });

        expect(wrapper.classes()).toContain('ik-input');
        expect(wrapper.classes()).toContain('ik-input--append');
        expect(wrapper.text()).toContain('Unit test...');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with prepend icon', () => {
        const wrapper = getInstance({
            props: {
                prepend_icon: 'times:light',
            },
        });

        expect(wrapper.classes()).toContain('ik-input');
        expect(wrapper.classes()).toContain('ik-input--prepend');
        expect(wrapper.findComponent(IkIcon).props('icon')).toBe('times:light');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with icon', () => {
        const wrapper = getInstance({
            props: {
                icon: 'times:light',
            },
        });

        expect(wrapper.classes()).toContain('ik-input');
        expect(wrapper.classes()).toContain('ik-input--prepend');
        expect(wrapper.findComponent(IkIcon).props('icon')).toBe('times:light');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with append icon', () => {
        const wrapper = getInstance({
            props: {
                append_icon: 'times:light',
            },
        });

        expect(wrapper.classes()).toContain('ik-input');
        expect(wrapper.classes()).toContain('ik-input--append');
        expect(wrapper.findComponent(IkIcon).props('icon')).toBe('times:light');
        expect(wrapper.element).toMatchSnapshot();
    });

    it('emits append and prepend click events', async () => {
        const prependClickFn = jest.fn();
        const appendClickFn = jest.fn();

        const wrapper = getInstance({
            props: {
                prepend_icon: 'times:light',
                append_icon: 'home',
            },
            attrs: {
                onIkClickPrepend: prependClickFn,
                onIkClickAppend: appendClickFn,
            },
        });

        await wrapper.findAllComponents(IkIcon).at(0)?.trigger('click');
        await wrapper.findAllComponents(IkIcon).at(1)?.trigger('click');

        expect(prependClickFn).toHaveBeenCalled();
        expect(appendClickFn).toHaveBeenCalled();
    });

    it('works with autofocus', async () => {
        const wrapper = getInstance({
            props: {
                autofocus: true,
            },
        });

        await wait(50);

        expect(wrapper.find('input').element).toBe(document.activeElement);
    });

    it('works with append_clear_icon', async () => {
        const wrapper = getInstance({
            props: {
                modelValue: '',
                append_clear_icon: true,
            },
        });

        expect(wrapper.findComponent(IkIcon).isVisible()).toBe(false);

        await wrapper.setProps({ modelValue: 'xyz' });

        expect(wrapper.findComponent(IkIcon).isVisible()).toBe(true);

        await wrapper.findComponent(IkIcon).trigger('click');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            [''],
        ]);
    });

    it('clamps number value to range', async () => {
        const wrapper = getInstance({
            props: {
                type: 'number',
                min: 0,
                max: 99,
            },
        });

        await wrapper.find('input').setValue('-1');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('0');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
            ['0'],
        ]);

        await wrapper.find('input').setValue('100');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
            ['0'],
            ['99'],
        ]);

        await wrapper.find('input').setValue('99.001');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
            ['0'],
            ['99'],
            ['99'],
        ]);
    });

    it('clamps number value to float range', async () => {
        const wrapper = getInstance({
            props: {
                type: 'number',
                min: '0.00',
                max: '99.77',
                decimals: 2,
            },
        });

        await wrapper.find('input').setValue('-1');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('0');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
            ['0'],
        ]);

        await wrapper.find('input').setValue('100');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
            ['0'],
            ['99.77'],
        ]);

        await wrapper.find('input').setValue('99.771');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
            ['0'],
            ['99.77'],
            ['99.77'],
        ]);

        await wrapper.find('input').setValue('99.77');

        expect(wrapper.emitted('update:modelValue')).toEqual([
            ['0'],
            ['0'],
            ['99.77'],
            ['99.77'],
            ['99.77'],
        ]);
    });

    it('works with simple decimal type', async () => {
        const wrapper = getInstance({
            props: {
                type: 'decimal',
            },
        });

        await wrapper.find('input').setValue('');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            [''],
        ]);

        await wrapper.find('input').setValue('0');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(2);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('0.0');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(3);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0.0'],
        ]);

        await wrapper.find('input').setValue('0,01');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(4);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0.01'],
        ]);

        await wrapper.find('input').setValue('-6,79');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(5);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['-6.79'],
        ]);

        await wrapper.find('input').setValue('  ');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(6);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('xxx');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(7);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('--7');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(8);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('-9.99.993');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(9);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['-999.993'],
        ]);

        await wrapper.find('input').setValue('-9,99.993');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(10);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['-999.993'],
        ]);

        await wrapper.find('input').setValue('-9,99.');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(11);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['-999.'],
        ]);
    });

    it('works with complex decimal type', async () => {
        const wrapper = getInstance({
            props: {
                type: 'decimal',
                min: 0,
                max: 13,
                decimals: 3,
            },
        });

        await wrapper.find('input').setValue('');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(1);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            [''],
        ]);

        await wrapper.find('input').setValue('0');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(2);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('0.0');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(3);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0.0'],
        ]);

        await wrapper.find('input').setValue('0,01');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(4);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0.01'],
        ]);

        await wrapper.find('input').setValue('-6,79');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(5);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('  ');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(6);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('xxx');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(7);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('--7');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(8);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['0'],
        ]);

        await wrapper.find('input').setValue('1.2.99311');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(9);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['12.99311'],
        ]);

        await wrapper.find('input').setValue('1,1,93');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(10);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['11.93'],
        ]);

        await wrapper.find('input').setValue('9,.');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(11);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['9.'],
        ]);

        await wrapper.find('input').setValue('99');
        expect(wrapper.emitted('update:modelValue')).toHaveLength(12);
        expect(wrapper.emitted('update:modelValue')?.slice(-1)).toEqual([
            ['13'],
        ]);
    });
});
