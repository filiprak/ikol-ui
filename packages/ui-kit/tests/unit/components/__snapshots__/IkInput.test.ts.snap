// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IkInput renders with append icon 1`] = `
<div
  class="ik-input ik-input--append"
>
  <!---->
  <input
    autocomplete="off"
    class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
    type="text"
    value=""
  />
  <div
    class="ik-input__append"
  >
    <i
      class="ik-icon ik-icon--clickable fa-light fa-times"
    />
  </div>
</div>
`;

exports[`IkInput renders with append slot 1`] = `
<div
  class="ik-input ik-input--append"
>
  <!---->
  <input
    autocomplete="off"
    class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
    type="text"
    value=""
  />
  <div
    class="ik-input__append"
  >
    
    Unit test...
    
  </div>
</div>
`;

exports[`IkInput renders with correct custom state 1`] = `
<div
  class="ik-input ik-input--disabled"
>
  <!---->
  <input
    autocomplete="on"
    class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
    disabled=""
    max="15"
    maxlength="100"
    min="0"
    minlength="4"
    name="unit-test"
    placeholder="Test placeholder..."
    readonly=""
    required=""
    step="55"
    type="password"
    value="Test value."
  />
  <!---->
</div>
`;

exports[`IkInput renders with correct default state 1`] = `
<div
  class="ik-input"
>
  <!---->
  <input
    autocomplete="off"
    class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
    type="text"
    value=""
  />
  <!---->
</div>
`;

exports[`IkInput renders with icon 1`] = `
<div
  class="ik-input ik-input--prepend"
>
  <div
    class="ik-input__prepend"
  >
    <i
      class="ik-icon fa-light fa-times"
    />
  </div>
  <input
    autocomplete="off"
    class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
    type="text"
    value=""
  />
  <!---->
</div>
`;

exports[`IkInput renders with prepend icon 1`] = `
<div
  class="ik-input ik-input--prepend"
>
  <div
    class="ik-input__prepend"
  >
    <i
      class="ik-icon fa-light fa-times"
    />
  </div>
  <input
    autocomplete="off"
    class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
    type="text"
    value=""
  />
  <!---->
</div>
`;

exports[`IkInput renders with prepend slot 1`] = `
<div
  class="ik-input ik-input--prepend"
>
  <div
    class="ik-input__prepend"
  >
    
    Unit test...
    
  </div>
  <input
    autocomplete="off"
    class="ik-input__native ik-f-input ik-f-input--size-sm ik-f-input--flat"
    type="text"
    value=""
  />
  <!---->
</div>
`;
